File: backend/app/parsers.py
# app/parsers.py
import os
from PyPDF2 import PdfReader
import docx

def extract_text_from_file(filepath: str, filename: str) -> str:
    ext = os.path.splitext(filename)[1].lower()

    if ext == ".txt":
        return _parse_txt(filepath)
    elif ext == ".pdf":
        return _parse_pdf(filepath)
    elif ext == ".docx":
        return _parse_docx(filepath)
    else:
        raise ValueError(f"Unsupported file type: {ext}")

def _parse_txt(filepath: str) -> str:
    with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
        return f.read()

def _parse_pdf(filepath: str) -> str:
    text = ""
    try:
        reader = PdfReader(filepath)
        for page in reader.pages:
            text += page.extract_text() or ""
    except Exception as e:
        raise ValueError(f"Error reading PDF: {e}")
    return text.strip()

def _parse_docx(filepath: str) -> str:
    text = ""
    try:
        doc = docx.Document(filepath)
        for para in doc.paragraphs:
            text += para.text + "\n"
    except Exception as e:
        raise ValueError(f"Error reading DOCX: {e}")
    return text.strip()

File: backend/utils/parsers.js
import fs from 'fs';
import path from 'path';
import pdfParse from 'pdf-parse';
import mammoth from 'mammoth';

export async function extractTextFromFile(filepath, filename) {
  const ext = path.extname(filename).toLowerCase();

  switch (ext) {
    case '.txt':
      return parseTxt(filepath);
    case '.pdf':
      return await parsePdf(filepath);
    case '.docx':
      return await parseDocx(filepath);
    default:
      throw new Error(`Unsupported file type: ${ext}`);
  }
}

function parseTxt(filepath) {
  try {
    const text = fs.readFileSync(filepath, 'utf-8');
    return text;
  } catch (error) {
    throw new Error(`Error reading TXT: ${error.message}`);
  }
}

async function parsePdf(filepath) {
  try {
    const fileBuffer = fs.readFileSync(filepath);
    const data = await pdfParse(fileBuffer);
    return data.text;
  } catch (error) {
    throw new Error(`Error reading PDF: ${error.message}`);
  }
}

async function parseDocx(filepath) {
  try {
    const result = await mammoth.extractRawText({ path: filepath });
    return result.value;
  } catch (error) {
    throw new Error(`Error reading DOCX: ${error.message}`);
  }
}
